SUS-262 | improve Wall performance

macbre opened the pull request:
https://wikia-inc.atlassian.net/browse/SUS-262

### Improvements implemented

* batch calls to `CommentsIndex::selectRow` to avoid multiple queries getting just a single row - introduce `CommentsIndex::newFromIds` (that is called just once per page request, instead of multiple times per a single message wall entry)
* cache the result of a first `ForumHelper::isForum` call - it's executed 30+ times for `$wgTitle`, hence the result is not likely to change :smile: 
* added in-memory cache for `WallMessage::newFromId` method that was called multiple times for the ID of the parent message wall posts
* improved the caching in `WallThread` (as suggested by @TK-999) and saved two additional DB queries (`CommentsIndex::newFromIds` and `Title::newFromIDs`)

### Results

I've tested my changes on a rather small message wall page (with just six messages). Number of queries made dropped from 233 to 77 (when visiting the page as an anon via `curl`).

### Comparing the performance

Let's observe the graphs of select queries performed on both `comments_index` and `page` table when handling Message Wall requests.

macbre committed:
ArticleComments / Wall: type hinting

macbre committed:
SUS-262 | introduce and use CommentsIndex::newFromIds

Instead of calling CommentsIndex::selectRow for each wall message post make a single call to CommentsIndex::newFromIds that will get them all in a batch

macbre committed:
ForumHelper::isForum - cache the result to avoid multiple calls to Title::newFromID

macbre committed:
WallMessage::newFromId - avoid two DB queries when called with $id = 0

Title::newFromId( 0 ) always returns null, WallMessage code was then making the same query to master DB

macbre committed:
WallMessage::newFromId | introduce in-memory cache

And log fallbacks to master when slave returns no results

macbre committed:
CommentsIndexTest: fix unit tests

macbre committed:
WallMessageTest: fix unit tests

rybmat commented on extensions/wikia/Wall/WallMessage.class.php:
```diff
@@ -39,24 +42,41 @@ function __construct( Title $title, $articleComment = null ) {
 		}
 
 		if ( $master == false ) {
+			// TODO: instead of relying on fallback to master let's implement a proper wfWaitForSlaves() use
```
what about this todo?

macbre replied:
Let's wait for the logs to go live, take a look at them and decide then. Maybe nothing will be logged (fingers crossed ;)).

rybmat reviewed (commented):
- do we need setDB method in CommentsIndex.class.php? looks like its unused

macbre commented:
@rybmat, `setDB` is called in `newFromId` method on a cloned instance of `CommentsIndex` object. Let's take a closer look...

macbre commented:
I'd rather not touch it at this point.

rybmat commented on extensions/wikia/Forum/ForumHelper.class.php:
```diff
@@ -2,15 +2,39 @@
-			} else if ( $wgTitle->getNamespace() === NS_USER_WALL_MESSAGE ) {
-				$mainTitle = Title::newFromId( $wgTitle->getText() );
+			} else if ( $title->getNamespace() === NS_USER_WALL_MESSAGE ) {
+				$mainTitle = Title::newFromID( $title->getText() );
```
is it correct?

rybmat replied:
per comment bellow it is...

macbre replied:
It's it, unfortunately :)

mszabo-wikia replied:
![It's by design](http://starecat.com/content/wp-content/uploads/bug-feature-by-design.jpg)

TK-999 commented:
Should we consider caching full `WallMessage` objects instead of just ids to spare lookup on `WikiActivity (as experimented in 1020474)?

mszabo-wikia reviewed (commented):
Should we consider caching full WallMessage objects instead of just ids to spare lookup on WikiActivity (as experimented in 1020474)?

macbre committed:
Merge branch 'dev' into SUS-262-improve-wall-performance

macbre commented:
@TK-999 - good point! I'm on it...

macbre committed:
WallThread: code cleanup + type hinting

macbre committed:
SUS-262: Save WallThread state to cache at this point, after objects have been initialized

As implemented by @TK-999 in 102047485ac6058cfc1e482b17dcc7719cde9b17

This saves two additional queries on cache hit:

      1 Title::newFromIDs
      1 CommentsIndex::newFromIds

mszabo-wikia reviewed (approved):
![LGTM](https://img.444.hu/hende_csaba.jpg)

macbre committed:
Merge branch 'dev' into SUS-262-improve-wall-performance

macbre committed:
CommentsIndex::newFromIds - prevent "DatabaseBase::makeList: empty input" exception when provided with an empty array of IDs

macbre commented:
Travis fails with `npm ERR! shasum check failed for /tmp/npm-2801-302c8201/registry.npmjs.org/sauce-connect-launcher/-/sauce-connect-launcher-0.6.1.tgz` - does not seem to be related.
